File: ch12_the_sparse_extended_information_filter/src/python/estimate.py
from numpy import *
from scipy.linalg import solve

from pdb import set_trace as bp

def estimate(m0,xi,omega,m):
    n = (xi.size-3)//2

    if n < 100:

       m1 = solve(omega,xi)

    else:
       m1=m
       for i in range(0,m0.size):
           ind = 2*m0[i]+1

           m1[ind:ind+2] = solve(omega[ind:ind+2,ind:ind+2],xi[ind:ind+2]-omega[ind:ind+2,:].dot(m1)+omega[ind:ind+2,ind:ind+2].dot(m1[ind:ind+2]))
       m1[0:3] = solve(omega[0:3,0:3],xi[0:3]-omega[0:3,:].dot(m1)+omega[0:3,0:3].dot(m1[0:3]))
    return m1

File: ch12_the_sparse_extended_information_filter/src/julia/estimate.jl
function estimate(m₀,μ,ξ,Ω)
    n = (length(μ)-3)÷2

    if n < 100
        μ .= Ω \ ξ
    else
        @inbounds for i ∈ m₀
            ind = [2*i+2,2*i+3]
            μ[ind] = Ω[ind,ind] \ (ξ[ind]-Ω[ind,:]*μ+Ω[ind,ind]*μ[ind])
        end
        μ[1:3] = Ω[1:3,1:3] \ (ξ[1:3]-Ω[1:3,:]*μ+Ω[1:3,1:3]*μ[1:3])
    end
end
